web-server-doc/web-server/scribblings/launch.scrbl

```diff
@@ -95,14 +193,34 @@ from a given path:
                           [#:tcp@ tcp@ (unit/c (import) (export tcp^)) raw:tcp@]
                           [#:ips+ports ips+ports (listof (cons/c (or/c string? false/c) (listof listen-port-number?))) (list (cons #f (list 80)))]
                           [#:max-waiting max-waiting integer? 511]
-                          [#:initial-connection-timeout initial-connection-timeout integer? 60])
+                          [#:initial-connection-timeout initial-connection-timeout integer? 60]
+                          [#:request-read-timeout request-read-timeout integer? 60]
```

Comment:
I'm wondering why these are `integer?` rather than `exact-positive-integer?` (or `(or/c exact-nonnegative-integer? +inf.0)`). It see that `#:initial-connection-timeout` and `#:max-waiting` were already there and specified as `integer?`, though I wonder what they really do if given negative integers. Did you want `#:request-read-timeout` to be consistent with `#:initial-connection-timeout`?

I would consider changing all of the documented contracts to `exact-positive-integer?` (or whatever we name the contract in <https://github.com/racket/web-server/pull/81>), even if we have to actually accept any `integer?` in some cases for backward compatability.

Reply:
I think both have the `integer?` contract by mistake. Their actual contracts are `number?`, but I think they should both be non-negative reals. Same with `response-timeout` and `response-send-timeout` (which are currently `exact-positive-integer`s). So we probably need two contracts

```racket
(define nonnegative-limit/c (or/c exact-nonnegative-integer? +inf.0))
(define timeout/c (and/c real? (>=/c 0)))
```

a negative `#:initial-connection-timeout` will immediately timeout all incoming connections, so it's probably safe to change that to `timeout/c`. That said, I think `timeout/c` should allow `0` since that can be helpful for testing purposes.

Reply:
@Bogdanp I've basically done as you suggest here in <https://github.com/racket/web-server/pull/84>. It turns out that `max-waiting` is restricted to `exact-nonnegative-integer?` by `tcp-listen`: making that more permissive seems hard (if even possible in a cross-platform way), so I just consistently use `exact-nonnegative-integer?` for now. For `timeout/c`, `(>=/c 0)` already checks `real?`, so the `and/c` isn't needed.